app/ota_client.py

```diff
@@ -896,48 +887,149 @@ def _gen_regular_files(self, rootfs_dir, regulars_file, target_dir):
-                                logger.debug(f"boot file: {regular_inf.path}")
-                                self._gen_boot_dir_file(
-                                    rootfs_dir, target_dir, regular_inf, prev_inf, fl
-                                )
-                            else:
-                                # others
-                                logger.debug(f"no boot file: {regular_inf.path}")
-                                self._gen_regular_file(
-                                    rootfs_dir, target_dir, regular_inf, prev_inf, fl
-                                )
-                            prev_inf = regular_inf
-                    except Exception as e:
-                        logger.exception("regular files setup error!:")
-                        res = False
-                    finally:
-                        os.chdir(cwd)
-        shutil.copy(log_file, "./regulars_dl.log")
-        return res
+                        except Exception as e:
+                            ecb(e)
+                            break
+                    else:
+                        try:
+                            pool.apply_async(
+                                self._process_regular_file,
+                                (rootfs_dir, target_dir, rfile_inf),
+                                error_callback=ecb,
+                            )
+                        except:
+                            # the only exception will be catched is ValueError
+                            # caused by calling pool.apply_async when the pool terminated
+                            pass
+
+                # stop accepting new tasks
+                pool.close()
+                # wait for all tasks to complete
+                # not apply timeout currently
+                pool.join()
+
+            # check if any exception is triggered
+            if not ecb_queue.empty():
+                # if any exception being raised in any child processes,
+                # raise it again in the main process.
+                logger.error(
+                    f"process regular files failed: {ecb_queue.get()}. All sub processess terminated."
+                )
+                raise OtaError(f"process regular files failed!")
+            else:  # everything is ALLRIGHT!
+                # update corresponding class attribute
+                self._process_regular_files_exit(gvar_dict)
+
+    def _process_regular_file(self, rootfs_dir, target_dir, rfile_inf: RegularInf):
+        """
+        main entry for paralleling processing regular files
+        """
+        try:
+            prev_inf = ""
+            # hardlinked file
+            if int(rfile_inf.links) >= 2:
+                # if the upcoming rfile entry is the first copy of hardlinked file
+                # then the prev_inf should be ""
+                prev_inf = global_var_dict["tmp-dict-hardlink_reg"].get(
+                    rfile_inf.sha256hash, ""
```

Comment:
prev_inf should be the RegularInf instance, so `rfile_inf` should be return instead of `rfile_inf.sha256`?

Reply:
Yes, the dict store <hash>:<RegularInf>.